### DISCUSSION TITLE
How can I get a future prediction of a cell in dynamo?

### DISCUSSION THREAD

**Discussion by chansigit** [aristoteleo/dynamo-release#413] [Q&A] [answered]
Dear authors,
Can you please tell me how should I implement the following requirements in dynamo?

Given each cell's expressional profile (x) and velocity (v), I hope to get predicted expression shortly after this moment (x + v Δt, Δt is a hyperparameter denoting a short period of time). I am developing a visualization method depending on this future prediction.

To my knowledge, `velocyto.R`'s function `gene.relative.velocity.estimates` can achieve this (https://www.rdocumentation.org/packages/velocyto.R/versions/0.6/topics/gene.relative.velocity.estimates). This function returns: "a list with velocity results, including the current normalized expression state ($current), projected ($projected) over a certain time ($deltaT) ..."

An intuitive solution I came up with is `min(0,   adata.X+adata.layers['velocity_S'] * delta_T)` (assuming adata.X is are counts). But I am not sure if it is valid. e.g. Is velocity_S  library-depth normalized? should I take unspliced read changes into consideration? 

I am looking forward to your answers. Thank you very much.

**Comment by Xiaojieqiu** [comment, marked as answer]
Dynamo learns a vector field so it can be used to predict over any arbitrary time and from any arbitrary initial states. At the core it peforms an integration of the vector field (integration of an ode function given some initial states, using scipy.integrate or scipy.integrate.solve_ivp) --- key ideas in numerical differential equation analyses. Please check the source code of our `dyn.pd.fate` (see below) function and the animations in our tutorial. 

Example on using fate: 
```
dyn.pd.fate(adata, basis='umap', init_cells=progenitor, interpolation_num=100,  direction='forward',
   inverse_transform=False, average=False, cores=3)
```

**Reply by chansigit** [reply]
Thank you Qiu. Velocity field is definitely a good solution. 

What I actually prefer is a rough solution without using the learned velocity field, namely a simple X+V*Δt form)  

It might sound weird to you. But let me explain. I am developing a method using denoised velocities to build the dynamo velocity field, which involves the current state X and a future prediction X+V*Δt. The `dyn.pd.fate` may lie downstream of the analysis in my design.

I am worried velocity_S is somehow normalized so that can not be directly added to the count-based adata.X. I tried to read the implementations in `sc.tl.dynamics` but it is really long and depressed me.

Can you please explain to me what unit/量纲 the dS/dt (velocity_S ) uses, or how dS/dt (velocity_S ) is normalized (or not)?

looking forward to your help ;D

**Reply by Xiaojieqiu** [reply]
sorry for the late reply (busy with another paper recently). 

Velocity_S is the V and the corresponding gene expression should be `M_s`. for splicing, dS/dt = u - \gamma S. Or dynamo's notation, it is Velocity_S = M_u - adata.var.gamma[:, None] M_s. 

Same as other tools, for splicing velocity, \beta is assumed to be 1 and gene expression is locally smoothed and gamma can be estimated via steady state assumption using either determinstic, moment or negative binomial distribution under steady state. 

See this function for more: https://github.com/aristoteleo/dynamo-release/blob/1d1f5c521d0b8763ab866dcc3c3d96be0ba3f8f3/dynamo/estimation/csc/velocity.py#L170

**Reply by chansigit** [reply]
Got it!  You are great.
